File: agents/tabos/remotesession.py
import asyncio
import json
import websockets

class RemoteSession:
	def __init__(self, local_session, sessionId):
		self.session = local_session
		self.agentNode = local_session.agentNode
		self.sessionId = sessionId
		self.ws = None
		self.isAlive=False
		self.is_connected = False
		self.execPms = None
		self.startPms=None
		self.wsTask=None
		self.msgTask=None
		self.messages=[]
		self.upperAgent = None
		self.askUpwardSeg = None
		self.fromAgent = None

	@staticmethod
	async def exec(session, node_name, agent, input_data, options=None):
		if options:
			startNodeOpts={"checkUpdate":options.get("checkUpdate",True)}
			callAgentOpts = {"fromAgent": options.get("fromAgent",None), "askUpwardSeg":options.get("askUpwardSeg",None)}
		else:
			startNodeOpts = {"checkUpdate": True}
			callAgentOpts = {}

		res = await session.callHub("StartAgentNode",{
			"name": node_name,
			"path": options.get("nodeEntry"),
			"options": startNodeOpts
		})
		if not res or res["code"] != 200:
			raise Exception(f"Start AgentNode error: {res.get('info', 'Unknown error')}")

		res= await session.callHub("AhCreateSession",{"node": node_name,"language":session.language})
		if not res or res["code"] != 200:
			raise Exception(f"Create Session error: {res.get('info', 'Unknown error')}")

		sessionId = res["sessionId"]
		remote_session = RemoteSession(session, sessionId)
		await remote_session.connect()
		await remote_session.start()
		return await remote_session.exec_agent(agent, input_data, callAgentOpts)

	# -----------------------------------------------------------------------
	async def connect(self):
		session=self.session
		ws = await websockets.connect(session.agentNode.host,max_size=10 * 1024 * 1024)
		self.ws = ws
		self.isAlive = True

	# -----------------------------------------------------------------------
	async def start(self):
		async def on_open():
			await self.ws.send(json.dumps({"msg": "CONNECT", "selector": self.sessionId}))
			print("Remote session WS connected.")

		self.startPms=asyncio.Future()
		try:
			self.wsTask=asyncio.create_task(self.runWS())
			self.msgTask = asyncio.create_task(self.runMsg())
			await on_open()
			await self.startPms
			# await asyncio.gather(on_open(), self.startPms)
		except Exception as e:
			print("RemoteSession start failed:", e)
			raise

	# -----------------------------------------------------------------------
	async def runMsg(self):
		async def on_message(msg):
			msg_vo = json.loads(msg)
			msg_code = msg_vo.get("msg")
			log=f"Message from remote session: {msg_vo}"
			if len(log)>2048:
				log=log[:2048]
			print(log)
			if not self.is_connected and msg_code == "CONNECTED":
				self.is_connected = True
				self.execPms = None
				self.startPms.set_result(True)
				self.startPms=None
				print("Remote session ready.")
			else:
				handler = getattr(self, f"WSMSG_{msg_code}", None)
				if handler:
					await handler(msg_vo)

		while self.isAlive:
			if len(self.messages)>0:
				msg=self.messages.pop(0)
				await on_message(msg)
			await asyncio.sleep(0.1)

	# -----------------------------------------------------------------------
	async def runWS(self):
		async def on_close():
			if self.is_connected:
				self.is_connected = False
				self.isAlive=False
				if self.execPms:
					self.execPms.set_exception(Exception("Session end."))
			else:
				if self.execPms:
					self.execPms.set_exception(Exception("Session start failed."))
			self.ws = None
			self.sessionId = None

		try:
			while self.isAlive:
				message = await self.ws.recv()
				if message:
					self.messages.append(message)
				await asyncio.sleep(0.1)
		except websockets.ConnectionClosed as e:
			print(f"Connection lost{e}")
			if self.isAlive and self.is_connected:
				await on_close()
		finally:
			if self.ws:
				await self.ws.close()

	async def endExec(self,result):
		self.isAlive=False
		self.is_connected=False
		await self.ws.close()
		await asyncio.gather(self.wsTask,self.msgTask)
		self.wsTask=None
		self.msgTask=None
		return result

	async def exec_agent(self, agent, prompt, opts):
		opts=opts or {}
		if not self.ws:
			raise Exception("RemoteSession missing websocket to execAgent")
		self.upperAgent = self.fromAgent = opts.get("fromAgent",None)
		self.askUpwardSeg = opts.get("askUpwardSeg",None)

		self.execPms = asyncio.Future()

		await self.ws.send(json.dumps({
			"msg": "ExecAgent",
			"sessionId": self.sessionId,
			"agent": agent,
			"prompt": prompt or ""
		}))

		try:
			result = await self.execPms
			print("RemoteSession execute result: ", result)
			await self.endExec(result)
			return result
		except Exception as e:
			print("Execution error:", e)
			raise e

	async def WSMSG_Message(self, msg):
		msg_code = msg["message"]["msg"]
		msg_vo = msg["message"]["vo"]
		session = self.session
		if msg_code == "ChatBlock":
			opts={}
			if msg_vo.get("image"):
				opts["image"] =msg_vo.get("image")
			if msg_vo.get("audio"):
				opts["audio"] =msg_vo.get("audio")
			role=msg_vo.get("role","assistant")
			await session.addChatText(role, msg_vo.get("text",""),opts)
		else:
			await session.sendToClient(msg_code, msg_vo)

	async def WSMSG_Call(self, msg):
		callId = msg.get("callId")
		msg_code = msg["message"]["msg"]
		msg_vo = msg["message"]["vo"]
		session = self.session
		try:
			if msg_code=="AskChatInput":
				result=await session.askChatInput(msg_vo)
			elif msg_code=="AskUpward":
				fakeAgent = {"upperAgent": self.fromAgent, "askUpwardSeg": self.askUpwardSeg}
				result = await session.askUpward(fakeAgent, msg_vo.get("prompt",None))
			elif msg_code=="AddWaitBlock":
				result=await session.showWait(msg_vo.get("text"))
			else:
				result=await session.callClient(msg_code, msg_vo)
			res={"msg": "CallResult", "sessionId": self.sessionId, "callId": callId, "result": result}
			await self.ws.send(json.dumps(res))
			return
		except Exception as e:
			errMsg={"msg": "CallResult", "sessionId": self.sessionId, "callId": callId, "error": f"{e}"}
			await self.ws.send(json.dumps(errMsg))

	async def WSMSG_ExecAgentEnd(self, msg):
		result = msg.get("result")
		error = msg.get("error")

		if error:
			if self.execPms:
				self.execPms.set_exception(Exception(error))
				self.execPms = None
		else:
			if self.execPms:
				self.execPms.set_result(result)
				self.execPms = None

File: agents/tabos/remotesession.mjs
import WebSocket, { WebSocketServer }from 'ws';

let RemoteSession,remoteSession;
//----------------------------------------------------------------------------
RemoteSession=function(localSession,sessionId){
	this.session=localSession;
	this.sessionId=sessionId;
	this.ws=null;
	this.isConnected=false;
	
	this.execCallback=null;
	this.execErrorCallback=null;
};
remoteSession=RemoteSession.prototype={};

//----------------------------------------------------------------------------
RemoteSession.exec=async function(session,nodeName,agent,input,options){
	let res,startNodeOpts,callAgentOpts;
	options=options||{};
	if(options){
		startNodeOpts={checkUpdate:options.checkUpdate??false};
		callAgentOpts={fromAgent:options.fromAgent??null,askUpwardSeg:options.askUpwardSeg??null};
	}else{
		startNodeOpts={checkUpdate:true};
		callAgentOpts={};
	}

	res=await session.callHub("StartAgentNode",{name:nodeName,path:options.nodeEntry,options:startNodeOpts,language:session.langauge});
	if(!res || res.code!==200){
		if(res){
			throw Error(`Start AgentNode error: ${res.info}`);
		}else{
			throw Error(`Start AgentNode error`);
		}
	}
	res=await session.callHub("AhCreateSession",{node:nodeName,language:session.language});
	if(!res || res.code!==200){
		if(res){
			throw Error(`Start AgentNode error: ${res.info}`);
		}else{
			throw Error(`Start AgentNode error`);
		}
	}
	let sessionId=res.sessionId;
	let remoteSsn=new RemoteSession(session,sessionId);
	await remoteSsn.start();
	return await remoteSsn.execAgent(agent,input,callAgentOpts);
};

//----------------------------------------------------------------------------
remoteSession.start=async function(){
	let pms,callback,callerror;
	pms=new Promise((resolve,reject)=>{
		callback=resolve;
		callerror=reject;
	});
	let selector,ws;	
	this.isConnected=false;
	selector=this.sessionId;
	ws=this.ws=new WebSocket(this.session.agentNode.host);
	ws.addEventListener('open',()=>{
		ws.send(JSON.stringify({msg:"CONNECT",selector:selector}));
		console.log("Remote session WS connected.");
	});
	ws.addEventListener('message', (msg) => {
		let msgVO,msgCode;
		console.log("Message from remote session:");
		console.log(msg)
		msgVO=JSON.parse(msg.data);
		msgCode=msgVO.msg;
		if(!this.isConnected){
			if(msgCode==="CONNECTED"){
				let call;
				this.isConnected=true;
				//Notify into chat session:
				call=callback;
				if(call){
					callback=null;
					callerror=null;
					call();
				}
				console.log("Remote session ready.");
			}else{
				let handler;
				handler="WSMSG_"+msgCode;
				handler=this[handler]||ws[handler];
				if(handler){
					handler.call(this,msgVO);
				}
				console.log("Message from remote session:");
				console.log()
			}
		}else{
			let handler;
			handler="WSMSG_"+msgCode;
			handler=this[handler]||ws[handler];
			if(handler){
				handler.call(this,msgVO);
			}
		}
	});
	ws.addEventListener('close', (msg) => {
		if(this.isConnected){
			let call;
			this.isConnected=false;
			call=this.execErrorCallback;
			if(call){
				this.execCallback=null;
				this.execErrorCallback=null;
				call("Session end.");
			}
			//Notify not connected...
		}else{
			let call;
			call=callerror;
			if(call){
				callback=null;
				callerror=null;
				call("Session start failed.");
			}
		}
		this.webSocket=null;
		this.sessionId=null;
	});
	ws.addEventListener('error', (msg) => {
		//wsError(msg);
		//TODO: Show error in session?
	});
	
	//Wait remote session ready:
	try{
		await pms;
	}catch(err){
		console.log("RemoteSession start failed:");
		console.error(err);
		return;
	}
};

//----------------------------------------------------------------------------
remoteSession.execAgent=async function(agent,prompt,opts){
	let ws,pms,callback,callerror,result;
	opts=opts||{};
	ws=this.ws;
	if(!ws){
		throw Error("RemoteSession missing websocket to execAgent");
	}
	pms=new Promise((resolve,reject)=>{
		this.execCallback=resolve;
		this.execErrorCallback=reject;
	});
	this.upperAgent=this.fromAgent=opts.fromAgent;
	this.askUpwardSeg=opts.askUpwardSeg;

	//Execute agent:
	ws.send(JSON.stringify({msg:"ExecAgent",sessionId:this.sessionId,agent:agent,prompt:prompt}));
	result=await pms;
	console.log("RemoteSession execute result: ");
	console.log(result);
	return result;
};

//----------------------------------------------------------------------------
//Agent send message to client:
remoteSession.WSMSG_Message=async function(msg){
	let session,callVO,msgCode,msgVO;
	callVO=msg.message;
	msgCode=callVO.msg;
	msgVO=callVO.vo;
	session=this.session;
	await session.sendToClient(msgCode, msgVO);
};

//----------------------------------------------------------------------------
//Agent call client:
remoteSession.WSMSG_Call=async function(msg){
	let session,callVO,callId,msgCode,msgVO,result;
	callId=msg.callId;
	callVO=msg.message;
	msgCode=callVO.msg;
	msgVO=callVO.vo;
	session=this.session;
	try{
		switch(msgCode){
			case "AskUpward": {
				result=await session.askUpward(this,msgVO.prompt);
				break;
			}
			default:{
				result =await session.callClient(msgCode,msgVO);
				break;
			}
		}
		this.ws.send(JSON.stringify({msg:"CallResult",sessionId:this.sessionId,callId:callId,result:result}));
		return result;
	}catch(err){
		let errMsg=JSON.stringify({msg:"CallResult",sessionId:this.sessionId,callId:callId,error:""+err});
		this.ws.send(errMsg);
	}
};

//----------------------------------------------------------------------------
//Agent send message to client:
remoteSession.WSMSG_ExecAgentEnd=async function(msg){
	let session,result,error,call;
	result=msg.result;
	error=msg.error;

	await this.ws.close(1000,"Finish");

	if(error){
		call=this.execErrorCallback;
		if(call){
			this.execCallback=null;
			this.execErrorCallback=null;
			call(error);
		}
	}else{
		call=this.execCallback;
		if(call){
			this.execCallback=null;
			this.execErrorCallback=null;
			call(result);
		}
	}
};

export default RemoteSession;
export {RemoteSession};
